presto-main/src/main/java/com/facebook/presto/operator/Driver.java

```diff
@@ -258,9 +258,10 @@ private void processNewSources()
         for (ScheduledSplit newSplit : newSplits) {
             Split split = newSplit.getSplit();
 
-            if (fragmentResultCacheContext.isPresent() && !(split.getConnectorSplit() instanceof RemoteSplit)) {
+            if (fragmentResultCacheContext.get().isPresent() && !(split.getConnectorSplit() instanceof RemoteSplit)) {
                 checkState(!this.cachedResult.get().isPresent());
-                this.cachedResult.set(fragmentResultCacheContext.get().getFragmentResultCacheManager().get(fragmentResultCacheContext.get().getHashedCanonicalPlanFragment(), split));
+                this.fragmentResultCacheContext.set(this.fragmentResultCacheContext.get().map(context -> context.removeRedundantSubfieldDomains(split.getConnectorSplit().getRedundantColumnDomains())));
```

Comment:
Is this correct? What if we have a running driver, the first split removes a column domain and this code will persist the change in the context, then the second split doesn't want to remove the column domain and it will get the one without column domain?

Reply:
Yes, this is correct. `fragmentResultCacheContext` in `Driver` is an `AtomicReference` to the initial context in `DriverContext`. Runtime changes will create a new `FragmentResultCacheContext` referenced by this `Driver`, but won't touch `DriverContext`. Given that one `Driver` would only handle one source split coming from storage, it should not affect other splits.